instrumentation/opentelemetry-instrumentation-urllib3/src/opentelemetry/instrumentation/urllib3/__init__.py

```diff
@@ -47,13 +47,21 @@ def strip_query_params(url: str) -> str:
 
 .. code:: python
 
-    # `request` is an instance of urllib3.connectionpool.HTTPConnectionPool
-    def request_hook(span, request):
```

Comment:
Note: the previous signature actually was: `def request_hook(span, instance, body)`

Reply:
I'm not too familiar with the inner workings of `urllib3` but does `HTTPConnectionPool` not have any of this information inside it (method, url, headers, body)? Would there be fields in it that would be useful, and if not, maybe we shouldn't even be including as part of the hooks. Most other instrumentations accept a `request` object of some sort which contains all information pertaining to the request itself.

Reply:
> I'm not too familiar with the inner workings of `urllib3` but does `HTTPConnectionPool` not have any of this information inside it (method, url, headers, body)?

Not really, [`HTTPConnectionPool`](https://urllib3.readthedocs.io/en/stable/reference/urllib3.connectionpool.html#urllib3.connectionpool.ConnectionPool) is just it, a connection pool. The instrumentation wraps [`urlopen`](https://urllib3.readthedocs.io/en/stable/reference/urllib3.connectionpool.html#urllib3.HTTPConnectionPool.urlopen) that accepts request parameters and uses a connection from the pool to send the request

> Would there be fields in it that would be useful, and if not, maybe we shouldn't even be including as part of the hooks

Indeed, it would make sense to me to remove the pool from the hook parameters. I just didn't want to do it without knowing why it's there in the first place, but I'm totally fine with removing it

> Most other instrumentations accept a `request` object of some sort which contains all information pertaining to the request itself

Indeed, it's just that the `urlopen` receives all this info as normal function parameters, and whoever made the initial implementation probably just forwarded them to the hook

How about this: I'd introduce our own `Request` class, put there the new `method` and `url`, and move the existing `headers` and `body` in there:

```python
@dataclass
class Request:
    method: str
    url: str
    headers: Mapping[str, str] | None
    body: bytes | IO[Any] | Iterable[bytes] | str | None
```

And then, the request hook becomes just `def request_hook(span: Span, request: Request)` with the possibility to extend `Request` in future without making it a breaking change. Thoughts?

Reply:
@eigenein we are doing something similar in httpx-instrumentation if you want take a look

Reply:
> And then, the request hook becomes just def request_hook(span: Span, request: Request) with the possibility to extend Request in future without making it a breaking change. Thoughts?

This sounds like a good approach to me. It would also be good to make the class exposed in the docs very clearly so users will know how to construct one fairly easily.